p4-16/psa/PSA.mdk

```diff
@@ -387,31 +408,36 @@ no payload.
 [INCLUDE=psa.p4:Action_ingress_truncate]
 ```
 
-### Clone/recirculation/resubmit
+### Clone, Recirculate, Resubmit
 
-~ Figure { #fig-clone; caption: "Clone/recirculate/resubmit in PSA"; page-align: here; }
+~ Figure { #fig-clone; caption: "Clone, recirculate, and resubmit in PSA"; page-align: here; }
 ![clone]
 ~
 [clone]: clone_recirc_resubmit.png { width: 90%; }
 
-Figure [#fig-clone] show two proposed architectures for clone/recirculate/resubmit in PSA.
+Figure [#fig-clone] shows the proposed architectures for clone,
+recirculate, and resubmit in PSA.
 
 #### Clone
 
-A PSA implementation provides a clone mechanism to create a copy of
-the original packet as an independent packet instance. The cloned
-packet can be a duplicate of the original packet or a copy of the
-deparsed packet after egress pipeline. The clone mechanism can submit
-the cloned packet to ingress parser or buffering mechanism. In a PSA
-implementation, the clone mechanism is implemented in PRE.
-
-The clone mechanism can optionally attach metadata to the cloned
+A PSA implementation provides a cloning mechanism to create copies of
+the original packet as independent packet instances. A cloned
+packet may be a duplicate of the original packet or a copy of the
+deparsed packet after egress pipeline. The cloning may mechanism submit
```

Comment:
Typo "cloning may mechanism".

Also @hanw said that he thought clone operations should be restricted to one of these possibilities only:

+ clone initiated at the end of the ingress control block can only be a clone of the original packet that triggered the clone, as received by the ingress parser, optionally with additional 'unordered metadata' (not headers) attached to it, currently specified via a clone.emit() method call in the ingress deparser (see Note 1 below).  The next place in the P4 code that will process such a clone will always be the egress parser.

+ clone initiated at the end of the egress control block, which can only be a clone of the packet as output by the egress deparser.  The next place in the P4 code that will process such a clone will always be the egress parser.

(Note 1) I believe that emit() is a misleading name to use for anything _except_ emitting headers, which have a specified bit order between the header and what gets put into the packet.  If the intent is that clone.emit() can be of a struct, and the data never leaves the PSA device on the wire, and therefore the fields might be packed in arbitrary order, and it will never be parsed by any parser in the PSA device, then I think it would be clearer to use a different method name, e.g. clone.attach_metadata() or something that conveys the behavior a bit better.